### PR TITLE
feat: invites regardless of registration type, invite privilege, groups to join on acceptance

### PR DESCRIPTION
**Pull request by gasoved**
Some notes:

Hidden groups are not included in the "groups to be joined" `<select>` to be consistent with the behavior in other places.

***

This leaves off joining of groups when the registration is completed by an approval (in approve-only registration mode), since it seems currently we can't distinguish a registration made through an invite link from one that is not. (Only mechanism for that right now is presence of an invite token.)  

In approval flow, invite keys are not deleted, and it is assumed the user is coming from invite link even when it is not (added to the "Invitations" table at /admin/manage/registration).  

In normal flow, invite-invitee "relationship" is deleted.  

We can have a follow-up on this (like an invite accepted "flag", and tangentially, showing groups from invites in the "Invitations" table).

### CODE REVIEW

File: src/views/modals/invite.tpl [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<div class="form-group">
+	<label for="invite-modal-emails">[[users:prompt-email]]</label>
+	<input id="invite-modal-emails" type="text" class="form-control" placeholder="friend1@email.net,friend2@email.net" />
```
Comment (julianlam) [inline src/views/modals/invite.tpl:3]: Let's use the traditional `example.com` here

File: public/src/admin/manage/users.js [hunk 2]
```diff
@@ -454,20 +454,57 @@ define('admin/manage/users', [
...
+				.filter((m, i, arr) => i === arr.indexOf(m))
+				.join(','),
+			groupsToJoin: $groups.val(),
+		};
+
+		if (!data.emails) {
+			return;
+		}
+
+		socket.emit('user.invite', data, function (err) {
```
Comment (julianlam) [inline public/src/admin/manage/users.js:503]: Let's rewrite this to use the Write API instead, instructions forthcoming.

File: public/src/client/users.js [hunk 3]
```diff
@@ -139,21 +139,59 @@ define('forum/users', [
...
+				.filter((m, i, arr) => i === arr.indexOf(m))
+				.join(','),
+			groupsToJoin: $groups.val(),
+		};
+
+		if (!data.emails) {
+			return;
+		}
+
+		socket.emit('user.invite', data, function (err) {
```
Comment (julianlam) [inline public/src/client/users.js:188]: Let's rewrite this to use the Write API instead, instructions forthcoming.

**Review by julianlam — CHANGES_REQUESTED** [review]

**Comment by julianlam** [comment]
Instead of introducing a new socket call, let's add a new route to the Write API.

Don't worry, this is a brand new feature that got merged last week, so it is not a surprise you did not know about it :smile: 

## Step 1
Add a new route to `src/routes/write/users.js`, let's go with `POST /invite`

## Step 2
Add the controller logic to `src/controllers/write/users.js`, `Users.inviteUser`

Note here that a lot of these controllers call out to `api` library. This is because we currently still support both socket.io and the write API, so the controller logic has to be organized in such a way that both methods can access.

In this case, this is a brand new route with no websocket variant (or rather, we're rewriting it to not be a socket call), so **you do not need to put your logic into the `api/users.js` file**. Just put it directly into this controllers file (`src/controllers/users.js`).

You'll note that the token generation/deletion controllers are done in this way. This is because there is no websocket equivalent.

## Step 3
Rewrite your call so that instead of socket.emit, you call it via:

```
require(['api'], (api) => {
    api.post('/users/invite', { email, groups })   // pseudocode
        .then(() => { // do things... });
});
```

The api module automatically handles errors by calling `app.alertError()`, but if you want to handle errors yourself, add a `.catch()`.

Slack me if you have questions!

**Comment by julianlam** [comment]
It is ok to use socket.io for `getInviteGroups`... I think. I'd prefer it to be a `GET` but perhaps we can refactor it out later.

File: src/routes/api.js [hunk 4]
```diff
@@ -36,4 +37,6 @@ module.exports = function (app, middleware, controllers) {
 	router.post('/topic/thumb/upload', middlewares, uploadsController.uploadThumb);
 
 	router.post('/user/:userslug/uploadpicture', middlewares.concat([middleware.exposeUid, middleware.authenticate, middleware.canViewUsers, middleware.checkAccountPermissions]), controllers.accounts.edit.uploadPicture);
+
+	helpers.setupApiRoute(router, 'get', '/user/invite-groups', [middleware.authenticate], controllers.user.getInviteGroups);
```
Comment (julianlam) [inline src/routes/api.js:41]: Let's move this to `src/routes/write/users.js`, `GET /api/v3/users/<uid>/inviteGroups`

Comment (gasoved) [reply src/routes/api.js]: This is used as a self-consumed endpoint, so {uid} route param seemed redundant. Change has been made nevertheless. :)

**Pull request merged**